### DISCUSSION TITLE
difference between LSP mode and "linter" mode; `from a import b` vs `import a.b`

### DISCUSSION THREAD

**Discussion by dkuettel** [microsoft/pyright#5000] [General]
Hello,

I know that tensorflow doesnt have terribly good type annotations yet. However, I seem to have some jumpy behavior with the LSP and it doesnt always seem to be the same as when I run it in "batch mode" using `pyright some_file.py`:

Generally, when in LSP mode, this doesnt work:

```python
from tensorflow import keras
keras.layers.Input  # None of this is completed
```

Yet, when I do:

```python
import tensorflow as tf
tf.keras.layers.Input  # everything gets completed beautifully
```

On top of that, the first version is shown as an error in LSP mode, but when I run it using `pyright some_file.py` I get no errors.

It's not completely clear that my LSP mode and "batch mode" are running in the exact same settings, but they do use the same `pyproject.toml`.

But as for the difference in LSP mode: Does the LSP mode have some kind of timeout how deep it scans into a library to inspect things? Is that maybe why it gets it only in the second case?

cheers, Daniel.

**Comment by hmc-cs-mdrissi** [comment]
I suspect this is because of "useLibraryCodeForTypes": true as default and amount of magical import weirdness tensorflow does. My guess is if you disable "useLibraryCodeForTypes": false it'll fail to complete/no errors.

"useLibraryCodeForTypes": true is helpful for IDE features, but is less reliable for type checking especially for libraries with a lot of dynamic magic. Tensorflow uses heavy amount of dynamic import magic so I've found "useLibraryCodeForTypes": true to be prone to false positives with it.

**Reply by erictraut** [reply]
The first code sample above appears to be a bug in pyright when `useLibraryCodeForTypes` is true. I'll investigate further.

You should see the same behavior between the LSP and the command-line version if you are using the same version of both. I recently (version 1.1.304) changed the default value of the `useLibraryCodeForTypes` from false to true. This was done to match the default value of pylance, which is the Microsoft's premium VS Code language server built on top of pyright.

> Does the LSP mode have some kind of timeout
No, we never use timeouts because that would result in non-deterministic behavior.

**Reply by dkuettel** [reply]
Ok in case it matters for you: I'm using tensorflow at `2.10.1` and my pyright is still `1.1.289`.

I'll check what difference it makes when I set `useLibraryCodeForTypes` explicitely. But the way I read the first reply is that it might not be a good setting when used for linting.

**Reply by dkuettel** [reply]
Ok just went thru the change list, maybe let me first double check if this problem persists with the newest pyright before you spend time on it. I'll report back as soon as I know.

**Reply by erictraut** [reply]
See https://github.com/microsoft/pyright/issues/5003.
